graph/generate.go

```diff
@@ -13,142 +9,177 @@ import (
-type msgContext struct {
-	msg        proto.Message
-	descendant Node
+// env represents the lexical scope a whistler message and its corresponding node belong to.
+// args is a list of the parent projector arguments.
+// argLookup is a map from whistler projector arguments to the nodes they generate.
+type env struct {
+	name      string
+	parent    *env
+	args      []proto.Message
+	argLookup map[proto.Message]Node
 }
 
-func (mc msgContext) String() string {
-	return fmt.Sprintf("msg: %v, descendent: %v", mc.msg, mc.descendant)
-}
-
-// dfsBuildStack is a stack adapted for building a lineage graph with a DFS algorithm.
-// In addition to pop and push, it has reversePush for maintaining target order.
-// It is not thread safe and needs to be updated for parallelized optimizations
-type dfsBuildStack struct {
-	stack []msgContext
-}
-
-func (s *dfsBuildStack) push(mc msgContext) {
-	if s.stack == nil {
-		s.stack = []msgContext{mc}
-	} else {
-		s.stack = append(s.stack, mc)
+func projectorEnv(parentEnv *env, projector *mbp.ProjectorDefinition, argMsgs []proto.Message, argLookup map[proto.Message]Node) *env {
+	return &env{
+		name:      projector.GetName(),
+		parent:    parentEnv,
+		args:      argMsgs,
+		argLookup: argLookup,
 	}
 }
 
-// reversePush pushes an array of proto messages onto the stack in reverse order. It assigns them all the same context.
-// The order is reversed to preserve the ordering of locally-defined whistler targets. For example, an array
-// of FieldMappings extracted from a Projector
-func (s *dfsBuildStack) reversePush(msgContexts []msgContext) {
-	for i := len(msgContexts) - 1; i >= 0; i-- {
-		s.push(msgContexts[i])
+// New uses a whistler MappingConfig to generate a new lineage graph.
+func New(mpc *mbp.MappingConfig) (Graph, error) {
+	projectors := make(map[string]*mbp.ProjectorDefinition)
+	for _, p := range mpc.GetProjector() {
+		projectors[p.GetName()] = p
 	}
-}
-
-func (s *dfsBuildStack) pop() (msgContext, bool) {
-	if s.len() <= 0 {
-		return msgContext{}, false
-	}
-	n := s.len() - 1
-	mc := s.stack[n]
-	s.stack = s.stack[:n]
-	return mc, true
-}
-
-func (s *dfsBuildStack) len() int {
-	fmt.Printf("getting length!")
-	return len(s.stack)
-}
-
-func (s *dfsBuildStack) String() string {
-	return fmt.Sprintf("%v", s.stack)
-}
 
-func makeMsgContexts(msgs []proto.Message, descendant Node) []msgContext {
-	msgContexts := make([]msgContext, len(msgs))
-	for i, msg := range msgs {
-		msgContexts[i] = msgContext{
-			msg:        msg,
-			descendant: descendant,
+	graph := Graph{
+		Edges:         map[IsID][]IsID{},
+		ArgumentEdges: map[IsID][]IsID{},
+		Nodes:         map[IsID]Node{},
+	}
+	e := &env{
+		name:      "root",
+		parent:    nil,
+		args:      []proto.Message{},
+		argLookup: map[proto.Message]Node{},
+	}
+	for _, mapping := range mpc.GetRootMapping() {
+		_, err := graph.addMsgLineage(e, mapping, nil, nil, projectors, false)
```

Comment:
nit: these two lines can be simplified as:

if _, err := ...; err != nil {
  return ...
}